sw/otbn/crypto/modexp.s

```diff
@@ -174,6 +179,399 @@ modexp:
 
   ret
 
+/**
+ * Constant-time bigint CRT modular exponentiation.
+ *
+ * Returns: C = modexp(A,E) = A^E mod M
+ *     where
+ *         E mod (p - 1) = E_p
+ *         E mod (q - 1) = E_q
+ *         M = p * q coprime.
+ *
+ * This implements the standard RSA CRT modexp algorithm, internally using the
+ * square and multiply via calls to `modexp`. Base modular exponentiations are
+ * performed in the Montgomery domain.
+ *
+ * The squared Montgomery modulus RR and the Montgomery constant m0' have to
+ * be precomputed and provided at the appropriate locations in dmem.
+ *
+ * Flags: The states of both FG0 and FG1 depend on intermediate values and are
+ *        not usable after return.
+ *
+ * The base bignum A is expected in the input buffer, the primes in the p and q
+ * buffers, the CRT components of the exponent E in the exp_p and exp_q buffers,
+ * and the CRT reconstruction coefficient in the q_p buffer. The result C is
+ * written to the output buffer.
+ *
+ * Note that the content of the exp_p and exp_q buffers will be modified during
+ * computation.
+ *
+ * @param[in]   x2: dptr_c, dmem pointer to buffer for output C
+ * @param[in]   x3: dptr_exp, work buffer for `modexp` calls
+ * @param[in]   x4: dptr_reduce, work buffer for modular reductions
+ * @param[in]  x17: dptr_m0d, dmem pointer to first limb of m0'
+ * @param[in]  x18: dptr_RR, dmem pointer to first limb of RR
+ * @param[in]  x23: dptr_a, dmem pointer to first limb of input A
+ * @param[in]  x25: dptr_e_p, dmem pointer to first limb of exponent CRT 
+       component E_p
+ * @param[in]  x26: dptr_e_q, dmem pointer to first limb of exponent CRT
+       component E_q
+ * @param[in]  x27: dptr_p, dmem pointer to first limb of cofactor p
+ * @param[in]  x28: dptr_q, dmem pointer to first limb of cofactor q
+ * @param[in]  x29: dptr_q_p, dmem pointer to first limb of CRT reconstruction
+       coefficient q_p
+ * @param[in]  x30: N, number of limbs per bignum
```

Comment:
nit: I think the cofactor limb count calculations assume n is divisible by 2, maybe it makes sense to add a note for that?
```suggestion
 * @param[in]  x30: N, number of limbs per bignum (must be even)
```